libraries/chain/combined_database.cpp

```diff
@@ -535,57 +541,280 @@ namespace eosio { namespace chain {
+         std::for_each (indices.begin(), indices.end(), add_row);
+         indices.clear();
+      }
+
+      void operator() (uint64_t contract, const char* key, std::size_t key_size,
+                     const char* value, std::size_t value_size){
+
+         b1::chain_kv::bytes composite_key(key, key + key_size);
+         auto [scope, table, remaining, type] = backing_store::db_key_value_format::get_prefix_thru_key_type(composite_key);
+
+         if (type == key_type::table) {
+            backing_store::payer_payload pp(value, value_size);
+
+            auto add_row = [this](const auto& row) { section_.add_row(row, db_); };
+
+            std::tuple<uint64_t, uint128_t, key256_t, float64_t, float128_t> secondary_key_types;
+            uint32_t total_count = primary_indices.size();
+            std::apply([this, &total_count](auto... index) {
+               total_count += ( this->secondary_indices<decltype(index)>().size() + ...); }, secondary_key_types);
+
+            add_row(table_id_object_view{account_name{contract}, scope, table, pp.payer, total_count});
+
+            add_row(fc::unsigned_int(primary_indices.size()));
+            std::for_each (primary_indices.begin(), primary_indices.end(), add_row);
+            primary_indices.clear();
+            std::apply([this](auto... index) { (this->write_secondary_indices(index),...); }, secondary_key_types);
+         } else if (type != key_type::primary_to_sec) {
+            std::invoke(extract_index_member_fun[ static_cast<int>(type) ], this, remaining, composite_key.end(), value, value_size);
+         }
+      }
+   };
+
+   template <typename Section>
+   const typename rocksdb_contract_table_writer<Section>::extract_index_member_fun_t rocksdb_contract_table_writer<Section>::extract_index_member_fun[] = {
+      &rocksdb_contract_table_writer<Section>::extract_primary_index,
+      nullptr,
+      &rocksdb_contract_table_writer<Section>::extract_secondary_index<uint64_t>,
+      &rocksdb_contract_table_writer<Section>::extract_secondary_index<uint128_t>,
+      &rocksdb_contract_table_writer<Section>::extract_secondary_index<key256_t>,
+      &rocksdb_contract_table_writer<Section>::extract_secondary_index<float64_t>,
+      &rocksdb_contract_table_writer<Section>::extract_secondary_index<float128_t>,
+   };
+
    void combined_database::add_contract_tables_to_snapshot(const snapshot_writer_ptr& snapshot) const {
       snapshot->write_section("contract_tables", [this](auto& section) {
-         index_utils<table_id_multi_index>::walk(db, [this, &section](const table_id_object& table_row) {
-            // add a row for the table
-            section.add_row(table_row, db);
+         if (kv_undo_stack && db.get<kv_db_config_object>().backing_store == backing_store_type::ROCKSDB) {
+            rocksdb_contract_table_writer<std::decay_t < decltype(section)>> writer(section, db, *kv_undo_stack);
+            auto undo = false;
+            if (kv_undo_stack->empty()) {
```

Comment:
Why not move this conditional and the following conditional to be within the walk_rocksdb_entries_with_prefix method?